### ISSUE TITLE
Swtich to Gitlab

### ISSUE THREAD

**Issue by theScrabi** [TeamNewPipe/NewPipe#1499]
ChoCho here comes the hype train.

Well. I'm not a big fan of __M$__ since they destroyed a lot of services they bought in the past, and I don't trust them at all. 

I don't purpose to switch away from github that quickly (maybe not even at all), but the way I'm used to how these companies work is they promise not to change something and F*** it up just a few months later. Just look at our beloved YouTube, and their all mighty awesome new subscription thing.
However weather we switch or not is up to you.

I would be fortified and prepare, so we can switch quickly once the time comes.

### [gitlab.com](gitlab.com)

| Pros | Cons |
| ----- | -----------------|
| - biggest alternative to github  | - veeery slow |
| - has a project importer that can import even the issues from github | - confusing/bad UI  |
| | - centralized (runs an Azure) |

### [git.schabi.org](git.schabi.org)

|| Pros | Cons |
| ----- | -----------------|
| - verry slick | - does not have an auto importer |
| - self hosted (runs on Hezner) | - does not support code review |
| | - github like UI |

## Preperation

So I think a step in the right direction for the beginning would be to try swtich our CI to a service that supports platforms other than github.com. Apropose would be [bitrise](https://www.bitrise.io/), or [gitlab CI](https://about.gitlab.com/features/gitlab-ci-cd/)

**Comment by Grammost** [comment]
First off, I also think that now since Microsoft bought Github, we will probably see that things change and very likely not for the better. 
I think your self-hosted Gittea will be better in the long-term, only the beginning will be quite difficult due to it not having an auto importer.

**Comment by pabru** [comment]
We could go for a mirrored approach, with git.schabi.org being the main site, and GitHub and GitLab being the mirrors. That way if any of them decide to take down the repo, the core is still within project's control. Other than having more redundancy, we also keep the benefit of people being able to discover and contribute to the code if they are coming from those more popular hosting platforms.

@theScrabi I've seen some interesting p2p git solutions appear recently. Maybe we could look into that as well?

**Comment by theScrabi** [comment]
> We could go for a mirrored approach

This is what I would suggest anyway. Wherever we go we have to mirror.

I also realized since we use [jitpac](jitpac.io) for the extractor we have to use a service that is supported by it.

**Comment by TotalCaesar659** [comment]
@theScrabi The table about GitLab is somewhat one-sided. You write that Gitlab is "centralized (runs an Azure)", but not about that it can be downloaded and deployed to your own server. I don't propose to get one, but it has to be mentioned too, I guess.
"confusing/bad UI" - it's a relative thing, tastes differ. There interfaces don't differ so much, you get used to it. Many large projects went to gitlab: [GNOME](https://www.gnome.org/news/2018/05/gnome-moves-to-gitlab-2/), [GIMP](https://www.gimp.org/news/2018/05/31/gimp-has-moved-to-gitlab/), [Inkscape](https://inkscape.org/en/news/2017/06/10/inkscape-moves-gitlab/) etc.
The only thing I can agree it's too slow. With its web-interface. But console git works too.

**Comment by theScrabi** [comment]
> The only thing I can agree it's too slow. With its web-interface. But console git works too.

This is the biggest disadvantage I have with it. Pourly not all things can be done via cli, and even pull push is slow.

**Comment by theScrabi** [comment]
Selfhoating gitlab was an alternative, since I witnesed that selfhosted it can get pretty fast, but then we still have the problem with jitpack

**Comment by Grammost** [comment]
@theScrabi Isn't (self-hosted) GitLab supported by Jitpack? Or are there other issues with that? https://jitpack.io/docs/PRIVATE/#self-hosted-git

**Comment by theScrabi** [comment]
I see, yea that would work :)

**Comment by SkyzohKey** [comment]
I use Self Hosted Gitlab CE on a day basis for work and I can assure you that it rocks waaaaaaay more than GH and isn't slower than it. The whole Git + CI + Deployments trio is a game changer to me and makes Gitlab way more elaborated & complete than GitHub for me. Give it a try you won't be deceived, trust me 😇

**Comment by theScrabi** [comment]
>  GH and isn't slower than it. 

Yea the selfhosted gitlab at our university is also not slow.

... Yea maybe we should consider selfhosted gitlab.  @TheAssassin what do you think.

**Comment by TheAssassin** [comment]
Folks, please calm down, and don't start to hastily switch from GitHub to some random alternatives. This doesn't make any sense, really. There's a lot of disadvantages in switching from GitHub to something else, see e.g., https://github.com/AppImage/AppImageKit/issues/803#issuecomment-396711385.

TL;DR: Before the end of 2018, it doesn't make _any_ sense to switch to some alternative. If the situation gets worse _then_, we can switch easily. I expect some migration helpers to be developed and become more mature by then, increasing the list of suitable alternatives. Then, one can discuss to switch to alternative platforms.

And please don't limit yourself to the two which are most GitHub-like. There's also self-hosted alternatives like e.g., [Phabricator](https://en.wikipedia.org/wiki/Phabricator), which employ a less repository but more project focused workflow, and provide a vast amount of tools that help non-devs to work more efficiently. It provides e.g., a mockup review tool that might come in handy for web development or UI discussions, and a simple non-public chat that can be used to send messages to people without knowing their mail address or having them join IRC. Sure, one has to get used to its workflow at first, but there must be a reason so many projects use it.

**Comment by vxcamiloxv** [comment]
+10

**Comment by pabru** [comment]
@TheAssassin A (partial) migration plan will be needed before then, to avoid chaos in case GitHub suddenly suspends or deletes the project. This is a real possibility for the NewPipe project. Microsoft is more likely to cave and suspend if a project is as much as _perceived_ to be "infringing copyright" or something similarly broad, certainly moreso than the previous management.
We shouldn't try to rush it, but we certainly should start shortlisting alternatives. I think that, whether we like it or not, we may not be able to remain on GitHub.

**Comment by TheAssassin** [comment]
@pabru we are already mirroring the repositories into several places for backup purposes, and what I am looking for currently is for someone to develop an issue export and migration tool, allowing us to make restorable backups of the issue trackers and migrate to any tool we want to once we need to actually migrate to some alternative platform. Ideally, that tool won't just create issues using some bot account, but allows to restore the issues properly.

Suggestions welcome!

**Comment by theScrabi** [comment]
@TheAssassin thats why i suggest gitlab they have an issue importer. I've tried it with somw of my other projects, and it works

**Comment by TheAssassin** [comment]
@theScrabi that is only half of what I required in my last comment. I'd like to have a way to actually back up the issues, and import them on demand, ideally on any platform.

**Comment by theScrabi** [comment]
True, but they don't support Inline code reviews.

**Comment by SkyzohKey** [comment]
All you request and more is available in Gitlab guys... Gitea is pretty cool project but it's still pretty young and not fully feature complete...

**Comment by TheAssassin** [comment]
@SkyzohKey no, it's not. That's the point. See https://github.com/TeamNewPipe/NewPipe/issues/1499#issuecomment-406281784. This "migration" stuff is far from good.

**Comment by TheAssassin** [comment]
> @TheAssassin this is going to be a vim vs. emacs discussion :D

That's wrong. I never claimed Gitea had that feature of "proper" migrations. But neither does GitLab. That's the problem with both of them.

**Comment by trymeouteh** [comment]
I am not a Microsoft fan either and I think by switching to gitlab will be a good move.

**Comment by Grammost** [comment]
As Github now blocks developers from certain countries, switching to an alternative is now more relevant than ever.

https://techcrunch.com/2019/07/29/github-ban-sanctioned-countries/

**Comment by TheAssassin** [comment]
So what? Where do you think GitLab is located? Running for the next platform doesn't make any difference, nor sense. It's not worth it.

**Comment by dimqua** [comment]
What about [notabug.org](https://notabug.org)?

**Comment by ChiefMikeK** [comment]
As stated in the tech crunch article
> A selection of GitHub services such as access to public repositories will remain available to everyone.

+ This is a public repo

**Comment by TheAssassin** [comment]
@dimqua next "big player".

If any, we'd go for self hosting, I guess. That's more sustainable and less annoying.

The big problem though is reaching developers. After all, on GitHub, it's fairly easy. On other platforms, the audience is much smaller. Noone likes to register at a ton of services.

**Comment by pandruszkow** [comment]
@TheAssassin GitLab supports [logging in with Google, Twitter, Facebook and GitHub](https://docs.gitlab.com/ce/administration/auth/README.html) , so maybe the users won't have to remember an extra username and password.

**Comment by theScrabi** [comment]
Self hosted? @TheAssassin eh? ... Then you could do mirrors to gitlab github git(you name it).
Or fediverse (don't know if that would affect the comunity)

**Comment by TomasGutierrez0** [comment]
> That's not my main concern. Not at all...

My bad, I worded it incorrectly, I made a few changes to my comment and left that line in there but I guess at least you could have explained your concerns if I didn't understand them, which I don't.

Is your concern related to being able to easily migrate information (issues, wiki, projects) from one git platform to another?

<hr>

On additional information about GitLab.
GitLab provides [free Ultimate/Gold](https://about.gitlab.com/solutions/open-source/program/) to open source projects, although GitLab core/free provides the same functionality as GitHub. I don't know how feasible is it to get it though.

**Comment by opusforlife2** [comment]
Can we confidently say NO to switching to just another USA-based company's repository after the youtube-dl DMCA fiasco?

**Comment by Andre601** [comment]
I personally don't think a switch is any good. People for once may be used to finding NewPipe on GitHub rather than any other site.
Not to mention the hundreds of open issues here that the people need to deal with.

And to give some info here: It has been now what? 2 years since Microsoft got GitHub? And what oh so horrible did happen so far? Pretty much nothing to really worry about.

Here are the imo good changes GitHub made since Microsoft aquired them:

- GitHub Actions became a thing (Useful for automation I would say)
- GitHub Discussions. Really useful for people who don't want to stick with IRC or similar.

Yes, the youtube-dl fiasco happened but only because of a faulty system where GitHub was just directly taking the repo down after a DMCA-claim to not get the risk of being fined a large sum, but they reinstated the repo since then and also made changes to their DMCA-System now to no longer let this happen that easily.

Just the fact that a large company aquires another one doesn't imediatally mean that it's doomed to die. I mean the picture of the OP clearly shows a view where people think Skype and Mojang are pretty much dead now, which couldn't be further from the truth (At least for the second one).

While the idea with the mirrors would be a nice compromise between things would this perhaps only cause more work for the maintainers as they would now have 2 if not more places to handle issues, PRs and more... And having everything handled on a central repo also would only be a half-good solution as people would require to have yet another account on a site for making one issue/PR if it isn't GitHub.

**Comment by TheEvilSkeleton** [comment]
I also recommend using [Codeberg](https://codeberg.org) if the GitLab UI is way too confusing. [SoundTouch](https://codeberg.org/soundtouch/soundtouch) recently switched to Codeberg from GitLab.

**Comment by Cyanic76** [comment]
What about [Codeberg](https://codeberg.org) ?

**Comment by Saarsk** [comment]
This isn't just about "MS bought Github, what will they do to it?" but also the fact that MS has a long history of being anti-competitive and being harmful to open-source, Linux etc. Just read about the history of how Winget came to be. With their AI going through code as well, it isn't a great thing for a FOSS project to be on Github.

I for one think the independence and sovereignty is a very important aspect. And support better gits instead of something owned by MS. You could say what you want about simply being anti-big corp, but there are a bunch of reasons hosting a project under a MS-product is not a great thing.

As for being able to reach the project hub, a simple redirect or "this project has moved to" will suffice. By pulling off the band-aid now it'll soon be like it was before instead of grudging the idea of initial change. If this was done when it was proposed it would be a new normal by now. The advantages for it triumph the downsides.

**Comment by theScrabi** [comment]
@Saarsk if we should ever move I agree: We should go to Codeberg instead of gitlab. If we move however should be up to the current maintainers. I see it like this: While you are quite right about your view on MS the takeover did not turn out to be too bad. Also we should never brake the ties to github compleatly as it is stile a good "display window" for opensource software.
